docs/rfcs/public-api-get-editor-for-audit-log.md

```diff
@@ -0,0 +1,238 @@
+the access token is got from internal environment, which is the service token 
+(https://github.com/bucketeer-io/bucketeer/blob/main/pkg/api/cmd/server.go#L114), so the email being shown 
+in audit log for public API write action is always internal Bucketeer email, which is not really correct.
+
+<div style="text-align: center;">
+  <img width="50%" src="/docs/rfcs/images/public-api-default-flow.png" alt="public api default flow">
+</div>
+
+We need to save the signature (in this case, email) of the API creator or maintainer and decide who should be the editor 
+then apply a mechanism to get the correct editor in api layer when making calls to the Web gRPC service.
+
+# Solutions
+
+## Problem 1: save the maintainer of API key when we create API key in Console
+
+We can add a column `maintainer` in api_key table so the maintainer can be saved. For old api keys, we can use creator
+in audit logs by filter type = 400 (create API key event) in audit_log table, select editor and `entity_id` 
+⇒ migration data for old `api_key` records and migration data for old `audig_log` records.
+
+Also, we need to hide the api key in the audit log snapshot, by adding a column `api_key` as new field to save api key
+and the old `id` column would be the UUID of the API key.
+
+<div style="text-align: center">
+  <img width="13%" src="/docs/rfcs/images/api-key-table-add-maintainer.png" alt="api_key table ERD" style="padding-right: 10px">
+  <img width="35%" src="/docs/rfcs/images/audit-log-snapshot.png" alt="audit log snapshot example">
+</div>
+
+**The message request of CreateAPIKey Web api should be:**
+
+```protobuf
+message CreateAPIKeyRequest {
+  string environment_namespace = 1;
+  string name = 2;
+  account.APIKey.Role role = 3;
+  string maintainer = 4;
+}
+```
+
+As we add new properties to `api_key` object, the response of the `GetAPIKeyBySearchingAllEnvironments` API 
+and `ListAPIKeys` API also return an extra `maintainer` and `id` field in `api_key` object:
+
+```protobuf
+message EnvironmentAPIKey {
+  string environment_namespace = 1 [deprecated = true];
+  APIKey api_key = 2;
+  bool environment_disabled = 3;
+  string project_id = 4 [deprecated = true];
+  environment.EnvironmentV2 environment = 5;
+  string project_url_code = 6;
+  string maintainer = 7;
+  string id = 8;
+}
+```
+
+## Problem 2: Get the correct editor in API layer when making call to Web service
+As we have saved API Key unique id, we can show it in the audit log like this (or just embed it in the api key name):
+```
+<key_name>-<id>-<creator_email> Feature flag has been created
+```
+I suggest 2 options:
+  - Let the maintainer be the creator of the API key.
+  - We let the client decide the editor of the action by adding editor email in the request body of public API.
+
+### Option 1: Let the editor be the creator of the API key <span style="color:green">(preferred)</span>
```

Comment:
Option 1 sounds good, but it would be even better to let the admin decide who will be the maintainer when creating the email instead of the creator.

Reply:
I see, so change the behavior of public write APIs won't be a problem?

Reply:
I didn't understand the reference.
Saving the creator's email automatically or letting the creator decide what email will be saved, what is the difference could affect the API?

Reply:
Oh I mean the code, not the behavior, sorry! Yeah if no affection then option 2 would work